src/main/kotlin/com/openlattice/indexing/BackgroundLinkingIndexingService.kt

```diff
@@ -155,22 +154,28 @@ class BackgroundLinkingIndexingService(
     }
 
     private fun mapCandidates(
-            candidates: List<Triple<UUID, OffsetDateTime, Set<UUID>>>
-    ): Pair<Map<UUID, Map<UUID, OffsetDateTime>>, Set<UUID>> {
+            candidates: List<Triple<List<String>, UUID, OffsetDateTime>>
+    ): Pair<Map<UUID, Map<UUID, Map<UUID, OffsetDateTime>>>, Set<UUID>> {
 
         // entity set id -> linking id -> last write
-        val linkingEntityKeyIdsWithLastWrite = mutableMapOf<UUID, MutableMap<UUID, OffsetDateTime>>()
+        val linkingEntityKeyIdsWithLastWrite = mutableMapOf<UUID, MutableMap<UUID, MutableMap<UUID, OffsetDateTime>>>()
         val linkingIds = mutableSetOf<UUID>()
         candidates.forEach {
-            val linkingId = it.first
-            val lastWrite = it.second
-            it.third.forEach { entitySetId ->
+            val linkingId = it.second
+            val lastWrite = it.third
+            it.first.forEach { rawEntityDataKey ->
+                val (entitySetId, originId) = rawEntityDataKey.split("|").map { UUID.fromString(it) }
+
                 if (!linkingEntityKeyIdsWithLastWrite.containsKey(entitySetId)) {
                     linkingEntityKeyIdsWithLastWrite[entitySetId] = mutableMapOf()
                 }
-                linkingEntityKeyIdsWithLastWrite.getValue(entitySetId)[linkingId] = lastWrite
+
+                if (!linkingEntityKeyIdsWithLastWrite.getValue(entitySetId).containsKey(originId)) {
+                    linkingEntityKeyIdsWithLastWrite.getValue(entitySetId)[originId] = mutableMapOf()
```

Comment:
I feel like this could be simplified with `.getValue(entitySetId).getOrPut(originId) {mutableMapOf()}` but I could be wrong

Reply:
Oh, nice catch! Thanks!